File: core/src/com/mygdx/game/MyGdxGame.java
package com.mygdx.game;

import com.badlogic.gdx.ApplicationAdapter;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.InputAdapter;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;

import java.util.ArrayList;
import java.util.List;

public class MyGdxGame extends ApplicationAdapter {
  SpriteBatch batch;
  Texture img;
  BoxObject box;
  List<CollidableObject> objects;
  int width;
  int height;

  Vector2 touchCoords = new Vector2(0, 0);

  @Override public void create() {
    batch = new SpriteBatch();
    width = Gdx.graphics.getWidth();
    height = Gdx.graphics.getHeight();
    img = new Texture("pengu.png");
    box = new BoxObject(new Rectangle(0,0,img.getWidth(), img.getHeight()), img, new Vector2(10, 10));
    box.setSpriteSize(img.getWidth(), img.getHeight());
    objects = new ArrayList<>();
    objects.add(box);
    objects.add(new CollidableObject(new Rectangle(-50,-50,50,height+100)));
    objects.add(new CollidableObject(new Rectangle(-50,height,width + 50,50)));
    objects.add(new CollidableObject(new Rectangle(width ,0,50,height+50)));
    objects.add(new CollidableObject(new Rectangle(-50,-50,width + 100,50)));

    Gdx.input.setInputProcessor(new InputAdapter() {
      @Override public boolean touchDown(int screenX, int screenY, int pointer, int button) {

        touchCoords = new Vector2(screenX, Gdx.graphics.getHeight() - screenY);

        if (touchCoords.x != 0 && touchCoords.y != 0) {
          for (int i = 0; i < objects.size(); i++) {
            BoxObject box;
            if (objects.get(i) instanceof BoxObject) {
              box = (BoxObject) objects.get(i);
            } else {
              continue;
            }
            if (box.isInside(touchCoords) && !box.isTouched) {
              Rectangle newBounds1 = new Rectangle(box.bounds.x - box.bounds.getWidth()/1.2f - 1, box.bounds.y - box.bounds.getHeight()/1.2f - 1, box.bounds.getWidth()/1.2f, box.bounds.getHeight()/1.2f);
              BoxObject newBox =
                  new BoxObject(new Rectangle(newBounds1), img,
                      new Vector2(-box.movement.x, -box.movement.y));
              newBox.setSpriteSize(box.sprite.getWidth() / 1.2f, box.sprite.getHeight() / 1.2f);
              objects.add(newBox);

              Rectangle newBounds2 = new Rectangle(box.bounds.x + box.bounds.getWidth()/1.2f + 1, box.bounds.y + box.bounds.getHeight()/1.2f + 1, box.bounds.getWidth()/1.2f, box.bounds.getHeight()/1.2f);
              BoxObject newBox2 =
                  new BoxObject(new Rectangle(newBounds2), img,
                      new Vector2(box.movement.x, box.movement.y));
              newBox2.setSpriteSize(box.sprite.getWidth() / 1.2f, box.sprite.getHeight() / 1.2f);
              objects.add(newBox2);

              newBox.isTouched = true;
              newBox2.isTouched = true;
              objects.remove(i);
            }
          }
        }
        return true;
      }
    });
  }

  @Override public void render() {

    Gdx.gl.glClearColor(0.17f, 0.85f, 0.68f, 1);
    Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
    batch.begin();

    for (int i = 0; i < objects.size(); i++) {
      if (objects.get(i) instanceof BoxObject) {
        ((BoxObject) objects.get(i)).render(batch, objects);
      }
    }
    batch.end();
    touchCoords.x = 0;
    touchCoords.y = 0;
  }

  @Override public void dispose() {
    batch.dispose();
    img.dispose();
  }
}

File: core/src/com/mygdx/game/MyGdxGame.kt
package com.mygdx.game

import com.badlogic.gdx.ApplicationAdapter
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.InputAdapter
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2

import java.util.ArrayList

class MyGdxGame : ApplicationAdapter() {
  lateinit var batch: SpriteBatch
  lateinit var img: Texture
  lateinit var player: BoxObject
  internal var objects: MutableList<CollidableObject> = mutableListOf()
  val MOVEMENT_SPEED = 5
  internal var width: Int = 0
  internal var height: Int = 0
  var keysPressed = mutableMapOf(19 to false, 20 to false, 21 to false, 22 to false)

  internal var touchCoords = Vector2(0f, 0f)

  override fun create() {
    batch = SpriteBatch()
    img = Texture("pengu.png")
    width = Gdx.graphics.width
    height = Gdx.graphics.height

    player = BoxObject(
        Rectangle(0f,0f, img.width.toFloat(), img.height.toFloat()), img, Vector2(10f, 10f)
    )
    player.setSpriteSize(img.width.toFloat(), img.height.toFloat())

    objects.add(player)

    objects.add(CollidableObject(Rectangle(-500f, -500f, 500f, (height + 500).toFloat())))
    objects.add(CollidableObject(Rectangle(-500f, height.toFloat(), (width + 500).toFloat(), 500f)))
    objects.add(CollidableObject(Rectangle(width.toFloat(), 0f, 500f, (height + 500).toFloat())))
    objects.add(CollidableObject(Rectangle(-500f, -500f, (width + 500).toFloat(), 500f)))

    Gdx.input.inputProcessor = object : InputAdapter() {
      override fun keyDown(keycode: Int): Boolean {
        keysPressed[keycode] = true
        println(keycode)

        return super.keyDown(keycode)
      }

      override fun keyUp(keycode: Int): Boolean {
        keysPressed[keycode] = false
        return super.keyUp(keycode)
      }

      override fun touchDown(
        screenX: Int,
        screenY: Int,
        pointer: Int,
        button: Int
      ): Boolean {

        touchCoords = Vector2(screenX.toFloat(), (Gdx.graphics.height - screenY).toFloat())

        if (touchCoords.x != 0f && touchCoords.y != 0f) {
          for (i in objects.indices) {
            val box: BoxObject
            if (objects[i] is BoxObject) {
              box = objects[i] as BoxObject
            } else {
              continue
            }
            if (box.isInside(touchCoords) && !box.isTouched) {
              val newBounds1 = Rectangle(
                  box.bounds.x +box.movement.x,
                  box.bounds.y +box.movement.y, box.bounds.getWidth() / 1.2f,
                  box.bounds.getHeight() / 1.2f
              )
              val newBox = BoxObject(
                  Rectangle(newBounds1), img,
                  Vector2(-box.movement.x, -box.movement.y)
              )
              newBox.setSpriteSize(box.sprite.width / 1.2f, box.sprite.height / 1.2f)
              objects.add(newBox)

              val newBounds2 = Rectangle(
                  box.bounds.x +box.movement.x,
                  box.bounds.y + +box.movement.y, box.bounds.getWidth() / 1.2f,
                  box.bounds.getHeight() / 1.2f
              )
              val newBox2 = BoxObject(
                  Rectangle(newBounds2), img,
                  Vector2(box.movement.x, box.movement.y)
              )
              newBox2.setSpriteSize(box.sprite.width / 1.2f, box.sprite.height / 1.2f)
              objects.add(newBox2)

              newBox.isTouched = true
              newBox2.isTouched = true
              objects.removeAt(i)
            }
          }
        }
        return true
      }
    }
  }

  override fun render() {

    Gdx.gl.glClearColor(0.17f, 0.85f, 0.68f, 1f)
    Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
    batch.begin()

    for (pair in keysPressed)
    {
      if(pair.value)
      {
        when(pair.key)
        {
          19, 51 -> player.bounds.y+=MOVEMENT_SPEED
          20, 47 -> player.bounds.y-=MOVEMENT_SPEED
          21, 29 ->player.bounds.x-=MOVEMENT_SPEED
          22, 32 -> player.bounds.x+=MOVEMENT_SPEED
        }

      }
    }

    for (i in objects.indices) {
      if (objects[i] is BoxObject) {
        (objects[i] as BoxObject).render(batch, objects)
      }
    }
    batch.end()
    touchCoords.x = 0f
    touchCoords.y = 0f
  }

  override fun dispose() {
    batch.dispose()
    img.dispose()
  }
}
